src/engraving/rendering/dev/chordlayout.cpp

```diff
@@ -2640,13 +2572,113 @@ void ChordLayout::layoutChords3(const MStyle& style, const std::vector<Chord*>&
     }
+
+    // Calculate the chords' dotPosX, and find the leftmost point for accidental layout
+    std::vector<Chord*> chordsWithNoVisibleDots;
+
+    for (Chord* chord : chords) {
+        bool _up = chord->up();
+
+        if (chord->stemSlash()) {
+            TLayout::layoutStemSlash(chord->stemSlash(), chord->stemSlash()->mutldata(), ctx.conf());
+        }
+
+        double overlapMirror;
+        Stem* stem = chord->stem();
+        if (stem) {
+            overlapMirror = stem->lineWidth() * chord->mag();
+        } else if (chord->durationType().headType() == NoteHeadType::HEAD_WHOLE) {
+            overlapMirror = style.styleMM(Sid::stemWidth) * chord->mag();
+        } else {
+            overlapMirror = 0.0;
+        }
+
+        double minDotPosX = 0.0;
+
+        std::vector<Note*> chordNotes = chord->notes();
+        std::sort(chordNotes.begin(), chordNotes.end(),
+                  [](Note* n1, const Note* n2) ->bool { return n1->line() <= n2->line(); });
+        for (Note* note : chordNotes) {
+            double noteX = 0.0;
+            if (note->ldata()->mirror()) {
+                if (_up) {
+                    noteX = chord->stemPosX() - overlapMirror;
+                } else {
+                    noteX = -note->headBodyWidth() + overlapMirror;
+                }
+            } else if (_up) {
+                noteX = chord->stemPosX() - note->headBodyWidth();
+            }
+
+            double ny = (note->line() + stepOffset) * stepDistance;
+            if (note->ldata()->pos().y() != ny) {
+                note->mutldata()->setPosY(ny);
+                if (stem) {
+                    TLayout::layoutStem(chord->stem(), chord->stem()->mutldata(), ctx.conf());
+                    if (chord->hook()) {
+                        chord->hook()->mutldata()->setPosY(chord->stem()->flagPosition().y());
+                    }
+                }
+            }
+            note->mutldata()->setPosX(noteX);
+
+            // find leftmost non-mirrored note to set as X origin for accidental layout
+            // a mirrored note that extends to left of segment X origin
+            // will displace accidentals only if there is conflict
+            double segmentX = noteX + chord->x();     // segment-relative X position of note
+            if (note->ldata()->mirror() && !_up && segmentX < -note->headBodyWidth() / 2) {
+                leftNotes.push_back(note);
+            } else if (segmentX < lx) {
+                lx = segmentX;
+            }
+
+            minDotPosX = std::max(minDotPosX, noteX + note->headBodyWidth() + chord->pos().x());
+        }
+
+        //---------------------------------------------------
+        //    layout dots simply
+        //     we will check for conflicts after all the notes have been processed
+        //---------------------------------------------------
+
+        if (chord->dots()) {
+            bool hasNoVisibleDots = true;
+            bool chordAlreadyHasDots = false;
```

Comment:
I think this should be a method of the Chord class. Maybe something like
```
bool Chord::hasDotsAllInvisible() const {
    if (!dots()) {
        return false;
    }

    for (Note* note : notes()) {
        for (Dot* dot : note->dots()) {
            if (dot->visible())
                return false;
        }
    }

    return true;
}
```

Reply:
Perhaps making it a static function in chordlayout.cpp is even better, in order not to clutter the Chord class with "single-use" methods.